crates/oxc_codegen/src/sourcemap_builder.rs

```diff
@@ -131,12 +136,37 @@ impl SourcemapBuilder {
 
     #[allow(clippy::cast_possible_truncation)]
     fn search_original_line_and_column(&mut self, position: u32) -> (u32, u32) {
-        let result = self
-            .line_offset_tables
-            .lines
-            .partition_point(|table| table.byte_offset_to_start_of_line <= position);
-        let original_line = if result > 0 { result - 1 } else { 0 };
-        let line = &self.line_offset_tables.lines[original_line];
+        let lines = &self.line_offset_tables.lines;
+        let mut idx = self.last_line_lookup;
+
+        // This is a hot path. When building sourcemaps, typically, positions are accessed in increasing order.
+        // e.g. if the last call to this function looked at position `n`, the next call will likely be for `n+1`.
+        if position >= lines[idx].byte_offset_to_start_of_line {
+            let cap = (idx + 16).min(lines.len() - 1);
+            while idx + 1 < cap && lines[idx + 1].byte_offset_to_start_of_line <= position {
+                idx += 1;
+            }
+            if lines[idx].byte_offset_to_start_of_line <= position {
+                idx = lines
+                    .partition_point(|table| table.byte_offset_to_start_of_line <= position)
+                    .saturating_sub(1);
+            }
```

Comment:
I think the reason for the perf gain getting erased is a bug here.

If I understand right, `lines[idx].byte_offset_to_start_of_line <= position` is always `true`, so it always ends up doing binary search, even if linear search already found the answer.

Ideally you want separate paths for "ended loop because hit cap" and "ended loop because we have the answer". It'd be easier to implement as a separate function from which you can early return.

Reply:
fixed. i think this ok for now? mnight be able to improve down the line

Reply:
Perf gain is gone again! There's a bug.

I've pushed some commits. All are pure refactor, just to make the code easier to follow. Now the bug is easier to spot.

Note on the last commit: I've changed `x.min(y)` to `min(x, y)`. I always find `x.min(y)` syntax very confusing and I always make mistakes when I use it (I end up using `min` when I mean `max`). I think you've fallen into that trap here too.